### PR TITLE
openai[patch]: ChatOpenAI.with_structured_output json_schema support

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/partners/openai/langchain_openai/chat_models/base.py [hunk 1]
```diff
@@ -298,6 +302,8 @@ class _AllReturnType(TypedDict):
 class BaseChatOpenAI(BaseChatModel):
     client: Any = Field(default=None, exclude=True)  #: :meta private:
     async_client: Any = Field(default=None, exclude=True)  #: :meta private:
+    root_client: Any = Field(default=None, exclude=True)  #: :meta private:
```
Comment (eyurtsev) [inline libs/partners/openai/langchain_openai/chat_models/base.py:305]: curious why is this necessary?

File: libs/partners/openai/langchain_openai/chat_models/base.py [hunk 2]
```diff
@@ -1339,6 +1422,20 @@ class AnswerWithJustification(BaseModel):
                 if is_pydantic_schema
                 else JsonOutputParser()
             )
+        elif method == "json_schema":
+            if schema is None:
+                raise ValueError(
+                    "schema must be specified when method is not 'json_mode'. "
+                    "Received None."
+                )
+            strict = strict if strict is not None else True
```
Comment (ccurme) [inline libs/partners/openai/langchain_openai/chat_models/base.py:1431]: to confirm: does `json_schema` with `strict=False` work?

**Review by ccurme — APPROVED** [review]
Looks great. Is the change to root poetry.lock related?

File: libs/partners/openai/langchain_openai/chat_models/base.py (hunk 2 above)
Comment (baskaryan) [reply libs/partners/openai/langchain_openai/chat_models/base.py:1431]: yep

**Comment by RicardodesaNunes** [comment]
In previous versions, I wanted to force the llm's response to be formatted as JSON. For that, I used the bind method to define the response format.
With the new update, I am getting an error that it cannot access `beta`.
Is there a way to solve this?

Here is the error:
```
File "/utils.py", line 114, in call_azure_openai
    res = llm.invoke(messages).content
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 5094, in invoke
    return self.bound.invoke(
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 284, in invoke
    self.generate_prompt(
  File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 756, in generate_prompt
    return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 613, in generate
    raise e
  File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 603, in generate
    self._generate_with_cache(
  File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 825, in _generate_with_cache
    result = self._generate(
             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/langchain_openai/chat_models/base.py", line 629, in _generate
    response = self.root_client.beta.chat.completions.parse(**payload)
               ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'beta'
```

Here is the code example I am using:
```
import langchain_openai 

LLM= langchain_openai.AzureChatOpenAI(
    openai_api_key=os.getenv("OPENAI_KEY"),
    azure_endpoint=os.getenv("OPENAI_ENDPOINT"),
    azure_deployment=os.getenv("OPENAI_DEPLOYMENT"),
    api_version="2024-02-01",
    model="gpt-4o"
)

def call_azure_openai(messages, json_output=True):
    if json_output:
        llm= LLM.bind(response_format={"type": "json_object"})
    res = llm.invoke(messages).content
    return json.loads(res) if json_output else res
```

**Comment by baskaryan** [comment]
> In previous versions, I wanted to force the llm's response to be formatted as JSON. For that, I used the bind method to define the response format. With the new update, I am getting an error that it cannot access `beta`. Is there a way to solve this?
> 
> Here is the error:
> 
> ```
> File "/utils.py", line 114, in call_azure_openai
>     res = llm.invoke(messages).content
>           ^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 5094, in invoke
>     return self.bound.invoke(
>            ^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 284, in invoke
>     self.generate_prompt(
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 756, in generate_prompt
>     return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 613, in generate
>     raise e
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 603, in generate
>     self._generate_with_cache(
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 825, in _generate_with_cache
>     result = self._generate(
>              ^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_openai/chat_models/base.py", line 629, in _generate
>     response = self.root_client.beta.chat.completions.parse(**payload)
>                ^^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'beta'
> ```
> 
> Here is the code example I am using:
> 
> ```
> import langchain_openai 
> 
> LLM= langchain_openai.AzureChatOpenAI(
>     openai_api_key=os.getenv("OPENAI_KEY"),
>     azure_endpoint=os.getenv("OPENAI_ENDPOINT"),
>     azure_deployment=os.getenv("OPENAI_DEPLOYMENT"),
>     api_version="2024-02-01",
>     model="gpt-4o"
> )
> 
> def call_azure_openai(messages, json_output=True):
>     if json_output:
>         llm= LLM.bind(response_format={"type": "json_object"})
>     res = llm.invoke(messages).content
>     return json.loads(res) if json_output else res
> ```

hm which version of the `openai` sdk do you have?

**Comment by ricnunespt** [comment]
checking the openai version I get:
```
Name: openai
Version: 1.40.3
```
I assumed the sdk version is obtained the same way as libraries versions. Feel free to correct me if there is another way.

**Comment by Dahimi** [comment]
Di 

> In previous versions, I wanted to force the llm's response to be formatted as JSON. For that, I used the bind method to define the response format. With the new update, I am getting an error that it cannot access `beta`. Is there a way to solve this?
> 
> Here is the error:
> 
> ```
> File "/utils.py", line 114, in call_azure_openai
>     res = llm.invoke(messages).content
>           ^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 5094, in invoke
>     return self.bound.invoke(
>            ^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 284, in invoke
>     self.generate_prompt(
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 756, in generate_prompt
>     return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 613, in generate
>     raise e
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 603, in generate
>     self._generate_with_cache(
>   File "/usr/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 825, in _generate_with_cache
>     result = self._generate(
>              ^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.11/site-packages/langchain_openai/chat_models/base.py", line 629, in _generate
>     response = self.root_client.beta.chat.completions.parse(**payload)
>                ^^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'beta'
> ```
> 
> Here is the code example I am using:
> 
> ```
> import langchain_openai 
> 
> LLM= langchain_openai.AzureChatOpenAI(
>     openai_api_key=os.getenv("OPENAI_KEY"),
>     azure_endpoint=os.getenv("OPENAI_ENDPOINT"),
>     azure_deployment=os.getenv("OPENAI_DEPLOYMENT"),
>     api_version="2024-02-01",
>     model="gpt-4o"
> )
> 
> def call_azure_openai(messages, json_output=True):
>     if json_output:
>         llm= LLM.bind(response_format={"type": "json_object"})
>     res = llm.invoke(messages).content
>     return json.loads(res) if json_output else res
> ```

I got the same error

**Comment by PaLoic1** [comment]
Hi @baskaryan, thanks for adding the support of JSON Schema to OpenAI, is planned to add it also to the AzureChat class ?

**Comment by baskaryan** [comment]
the above azure openai error and general azure support should be included in langchain-openai==0.1.22, would love to know if that resolves the issues / works for you @Dahimi @PaLoic1 @ricnunespt

**Comment by PaLoic1** [comment]
Thanks for your feedback, I check the code of `with_structured_output` and it seems does not support `json_schema` [file](https://github.com/langchain-ai/langchain/blob/master/libs/partners/openai/langchain_openai/chat_models/azure.py#L680) like in the `ChatOpenAI` class [here](https://github.com/langchain-ai/langchain/blob/1aa2e5706f0e60390f0fe5631b3517bce534812d/libs/partners/openai/langchain_openai/chat_models/base.py)`

**Comment by baskaryan** [comment]
you're right, fixing here https://github.com/langchain-ai/langchain/pull/25591
